Ast10x0 cache

aspeeddylan opened the pull request:
Add cache driver and enable cache for Aspeed AST10x0 series SOC.  Aspeed AST10x0 series SOC has a embedded Cortex-M3 CPU which doesn't have native cache memory.  Aspeed implements a integrated system level cache for instruction and data accesses.

aspeeddylan committed:
cache: Add Aspeed AST10x0 external cache driver

The CPU in Aspeed AST10x0 SOC is a ARM Cortex-M4 which doesn't internal
cache memory.  Aspeed implements an integrated system level cache to
accelerate instruction and data memory accesses.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>

aspeeddylan committed:
soc: arm: select CACHE_ASPEED for Aspeed AST10x0 series SOC

Select CACHE_ASPEED to enable Aspeed cache driver for AST10x0 series
SOC.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>

aspeeddylan committed:
soc: arm: aspeed: enable cache for AST10x0 series SOC

Enable cache for AST10x0 series SOC in platform initialization.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>

carlocaione commented on drivers/cache/Kconfig:
```diff
@@ -16,3 +16,4 @@ config CACHE_HAS_DRIVER
 	bool
 
 endif # CACHE
+source "drivers/cache/Kconfig.aspeed"
```
nit: add an empty line before

carlocaione commented on drivers/cache/Kconfig.aspeed:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021 ASPEED Technology Inc.
```
nit: 2022

carlocaione commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,267 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
```
`<zephyr/kernel.h>`

carlocaione commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,267 @@
+#define CACHE_INVALID_REG      0xa54
+#define CACHE_FUNC_CTRL_REG    0xa58
+
+#define CACHED_SRAM_ADDR       DT_REG_ADDR_BY_IDX(DT_NODELABEL(sram0), 0)
+#define CACHED_SRAM_SIZE       DT_REG_SIZE_BY_IDX(DT_NODELABEL(sram0), 0)
```
Can you just use `CONFIG_SRAM_BASE_ADDRESS` and `CONFIG_SRAM_SIZE` instead?

carlocaione commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,267 @@
+#define CACHED_SRAM_END        (CACHED_SRAM_ADDR + CACHED_SRAM_SIZE - 1)
+
+#define CACHE_AREA_SIZE_LOG2   15
+#define CACHE_AREA_SIZE         (1 << CACHE_AREA_SIZE_LOG2)
```
nit: tab is off by 1?

carlocaione commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,267 @@
+
+	ARG_UNUSED(op);
+
+	/* enter critical section */
+	if (!k_is_in_isr())
+		key = irq_lock();
+
+	if (((uint32_t)addr < CACHED_SRAM_ADDR) ||
+	    ((uint32_t)addr > CACHED_SRAM_END)) {
+		return 0;
+	}
```
if `addr` is wrong you are leaving the function with the IRQs locked.

carlocaione commented on soc/arm/aspeed/ast10x0/Kconfig.series:
```diff
@@ -12,5 +12,6 @@ config SOC_SERIES_AST10X0
 	select SOC_FAMILY_ASPEED
 	select CACHE
 	select CACHE_MANAGEMENT
+	select CACHE_ASPEED
```
are you still missing `CONFIG_HAS_EXTERNAL_CACHE` somewhere though?

aspeeddylan replied:
I added in Kconfig.defconfg.series.  Should I move it to somewhere?
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/aspeed/ast10x0/Kconfig.defconfig.series#L18-L19

carlocaione replied:
interesting. I would have expected that change to be in this PR. Anyway, that's fine.

aspeeddylan commented:
Hi @carlocaione , 
Thank you for your review.  Should I modify my commits and re-push? Or it will be better if I make additional commits to fix the the codes you pointed out?

carlocaione commented:
> Hi @carlocaione , Thank you for your review. Should I modify my commits and re-push? Or it will be better if I make additional new commits to fix the the codes you pointed out?

You should fix your commits and force-push again.

carlocaione commented on drivers/cache/Kconfig.aspeed:
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 #
-# Copyright (c) 2021 ASPEED Technology Inc.
+# Copyright (c) 2022 ASPEED Technology Inc.
```
this change should be squashed in the first commit

carlocaione commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,267 @@
+{
+	uint32_t base = DT_REG_ADDR(DT_NODELABEL(syscon));
+
+	sys_write32(0, base + CACHE_FUNC_CTRL_REG);
```
you cannot really use `sys_write32` & co to interface with syscon. We have in place a syscon API for that. Look at `include/zephyr/drivers/syscon.h`. You should use `syscon_get_base()`, `syscon_read_reg()` & co.

aspeeddylan replied:
Thank you for your comment.  I've added a commit to enable generic syscon driver and used syscon api to access syscon registers.

aspeeddylan committed:
dts: arm: aspeed: add compatible string for syscon

Bind generic syscon driver for syscon in Aspeed AST10x0 series SOC.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>

carlocaione commented on drivers/cache/Kconfig.aspeed:
```diff
@@ -0,0 +1,12 @@
+        bool "ASPEED external cache driver"
+        depends on SOC_SERIES_AST10X0
+        select CACHE_HAS_DRIVER
+        select SYSCON
+        select SYSCON_GENERIC
```
uhm, this should `depend` on syscon, and syscon should be enabled by defconfig (or some other Kconfig file) instead.

aspeeddylan replied:
Thank you for your comment.  I've update the commits.

aspeeddylan committed:
soc: arm: select SYSCON for Aspeed AST10x0 series SOC

Aspeed AST10x0 series SOC has a SYSCON hardware block at address
0x7e6e2000.  Enable this option to enable relative driver.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>

carlocaione commented on drivers/cache/Kconfig.aspeed:
```diff
@@ -0,0 +1,10 @@
+
+config CACHE_ASPEED
+bool "ASPEED external cache driver"
+	depends on SOC_SERIES_AST10X0
+	select CACHE_HAS_DRIVER
```
Sorry, maybe I was not clear enough. You should add here `depends on SYSCON` since your driver is actually depending on some syscon API.

aspeeddylan replied:
Got it.  Force push the modified commit.

povergoing commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,275 @@
+	/* calculate how many areas need to be set */
+	start_bit = CACHED_SRAM_ADDR >> CACHE_AREA_SIZE_LOG2;
+	end_bit = CACHED_SRAM_END >> CACHE_AREA_SIZE_LOG2;
+	syscon_write_reg(dev, CACHE_AREA_CTRL_REG, GENMASK(end_bit, start_bit));
```
Is it possible that the `end_bit `is too large that causing the `uint32_t ` overflow? does it matter in your system?

aspeeddylan replied:
Thank you for your comment.  The max RAM size in this SOC is 768KB so theoretically it wouldn't overflow.  But I guess I can add a ceil for ```start_bit``` and ```end_bit```.
 
 ````
 start_bit = MIN(31, CACHED_SRAM_ADDR >> CACHE_AREA_SIZE_LOG2);
 end_bit = MIN(31, CACHED_SRAM_END >> CACHE_AREA_SIZE_LOG2);
 ```

povergoing replied:
Got it. it's up to you, as long as it works for your system :)

aspeeddylan replied:
Re-pushed.  Thanks for your review.

povergoing commented on drivers/cache/cache_aspeed.c:
```diff
@@ -0,0 +1,275 @@
+	n = get_n_cacheline((uint32_t)addr, size, &aligned_addr);
+
+	for (i = 0; i < n; i++) {
+		syscon_write_reg(dev, CACHE_INVALID_REG, 0);
```
so it's kind of a hardware issue. thanks for your clarification.

aspeeddylan commented:
@carlocaione @povergoing sorry that somehow your previous approves are dismissed after I re-push my commits.  Could you kindly review again? Thank you.

povergoing reviewed (approved):
Refresh +1

aspeeddylan commented:
Could anyone teach me what is the next step to make this PR be merged?

carlocaione commented:
> Could anyone teach me what is the next step to make this PR be merged?

We are in code freeze. This will be merged when the merge window will be open again.
